### ISSUE TITLE
After uninstalling calico the nodes stuck in NetworkUnavailable because of CalicoIsDown

### ISSUE THREAD

**Issue by vasiliev-a-a** [projectcalico/calico#7816]
Hello. I am new to Kubernetes, so i explore various CNI plugins. Tried out calico - it works perfectly. But it seems like it cannot be removed cleanly. After i remove Installation resource and tigera-operator resources all my nodes stuck with condition NetworkUnavailable with reasons CalicoIsDown.
Tried to reset an re-join nodes - no result.

## Expected Behavior
I believe nodes should return into "No CNI plugin" state, and after valid configuration is provided become "Ready".

## Current Behavior
When calico and operator manifests are deleted with kubectl nodes stuck in NetworkUnavailable condition. Installing different CNI plugin does not fix the problem.

### 1. Deploying
```bash
kubectl create -f https://raw.githubusercontent.com/projectcalico/calico/v3.26.1/manifests/tigera-operator.yaml
```
```bash
kubectl apply -f - << EOF
apiVersion: operator.tigera.io/v1
kind: Installation
metadata:
  name: default
spec:
  calicoNetwork:
    ipPools:
    - blockSize: 28
      cidr: 10.10.0.0/22
      encapsulation: None
      natOutgoing: Disabled
      nodeSelector: all()
EOF
```

### 2. Observing success
```bash
kubectl get node -o jsonpath --template='{range .items[*]}{.metadata.name}{"\n"}{range .status.conditions[*]}{"\t"}{.type}{" : "}{.status}{" -> "}{.reason}{" ("}{.message}{")"}{"\n"}{end}{end}'
```
![image](https://github.com/projectcalico/calico/assets/115141812/21602e76-4ae6-4edc-98d0-0a074e552f3c)
```bash
kubectl get -n kube-system deployment
```
![image](https://github.com/projectcalico/calico/assets/115141812/616729d0-ef2a-4197-bd05-38141a3df77c)

### 3. Uninstalling
```bash
kubectl delete installation default
```
```bash
kubectl delete -f https://raw.githubusercontent.com/projectcalico/calico/v3.26.1/manifests/tigera-operator.yaml
```

### 4. Observing problem
*Provide a valid CNI config of your choice*
*Recreate coredns*
```bash
kubectl delete -n kube-system pod -l k8s-app=kube-dns
```
```bash
kubectl get -n kube-system pod -l k8s-app=kube-dns -o custom-columns=name:.metadata.name,status:.status.phase,message:.status.conditions[*].message
```
![image](https://github.com/projectcalico/calico/assets/115141812/7934d60e-864b-4c61-93a8-bd534d3b6ecd)
```bash
kubectl get node -o jsonpath --template='{range .items[*]}{.metadata.name}{"\n"}{range .status.conditions[*]}{"\t"}{.type}{" : "}{.status}{" -> "}{.reason}{" ("}{.message}{")"}{"\n"}{end}{end}'
```
![image](https://github.com/projectcalico/calico/assets/115141812/f41c5855-6fba-4321-b662-8436fee0558d)

### Note
If a toleration for 'node.kubernetes.io/network-unavailable' is applied on the deployment its pods are scheduled and got IPs, thus CNI is working fine. But calico causes a node condition of `NetworkUnavailable` and the respective taints prevent scheduling.
![image](https://github.com/projectcalico/calico/assets/115141812/f019be90-6ce6-4eff-91ed-70047002c170)

## Your Environment
Calico: v3.26.1
Kubernetes: v1.27.3
Debian GNU/Linux 11 (bullseye)
Kernel: 5.10.0-23-amd64
Containerd: 1.6.21

**Comment by caseydavenport** [comment]
This is a tricky one - Calico needs to know the difference between "Calico is being restarted / upgraded" and "Calico is being removed from the cluster", and of course once Calico is removed from the cluster, there is no longer any component present in the cluster to actually remove the condition!

In general, uninstalling a CNI plugin from a cluster is a destructive operation - there will be other issues you hit after this as well, such as artifacts left on the node (iptables, routes, CNI plugin files and configuration, etc) and any existing pods in the cluster will fail to function. The cleanup will vary from plugin to plugin, so I think the normal advice is to spin up a new cluster rather than try to re-use an existing one after uninstalling networking from that cluster.

That said, there may still be a way to make this particular behavior slightly nicer.

**Comment by caseydavenport** [comment]
> Is it possible to introduce some thing like calicoctl clean option to reset the conditions and taints?

This could work, yeah. It wouldn't clean up any of the programming on the nodes themselves - that would still require a node reboot at a minimum.

**Comment by boedy** [comment]
I'm in the same boat. I'm trying to remove the taint, but I'm not able to. Is there a way to manually remove the condition?

**Comment by dberardo-com** [comment]
i get an error with the script @boedy : curl: (58) unable to load client key: -8178 (SEC_ERROR_BAD_KEY)


any hints ?

**Comment by boedy** [comment]
@dberardo-com Did you run the command on one of the control plane nodes? This would only work on k3s clusters.

Check if `CLIENT_CERT_DATA` and `CLIENT_KEY_DATA` contain any data.

**Comment by dberardo-com** [comment]
they do ... so strange ..

also i seem to have found a not-so-nice workaround, which is basically to disable the calico operator and destroy all workloads related to calico apartfrom the "caloc-nodes" deployment. the calico nodes pods will then stuck permamently on a "pending/restart" loop because they dont find all other needed pods, but the nodes do not get tained as unschedulable ...

**Comment by qpanpony** [comment]
Successfully cleaned up the nodes in K8s using @boedy's script. Just replaced env variable K3S_CONFIG with env variable   K8S_ADMIN_CONFIG="/etc/kubernetes/admin.conf" and changed the following curl command's --cacert parameter into "/etc/kubernetes/pki/ca.crt".

Though as said by @caseydavenport, uninstalling a CNI plugin from a cluster is a destructive operation. It's still a bit awkward that Calico can not be removed gracefully from the cluster.

**Comment by dberardo-com** [comment]
@boedy  i can confirm that i am running the script on control plane of k3s  cluster and that those vars have content : https://github.com/projectcalico/calico/issues/7816#issuecomment-1714202942

i think this is related to the private key fromat: https://superuser.com/questions/1482345/how-to-use-an-ec-private-key-with-curllibnss

**Comment by caseydavenport** [comment]
> Right now i am using the kube-router, but have to apply node.kubernetes.io/network-unavailable toleration to all Pods

FWIW, I am not sure why kube-router isn't tolerating this already. Doesn't it need to tolerate unavailable network if it itself provides network to the cluster? Calico certainly does.

I would sort of expect any other implementation to tolerate this, and then to set network-unavailable to "false" once it is installed.

**Comment by vasiliev-a-a** [comment]
> I would sort of expect any other implementation to tolerate this, and then to set network-unavailable to "false" once it is installed.

That is a good point. May be i should open an issue there too.
I am also curious of how other CNI plugins, like Cilium or Weave Net, would behave 🤔

Nevertheless, i think that any solution, when it gracefully removed, should revert the system to a clean state, just like before it was installed.

**Comment by caseydavenport** [comment]
> Nevertheless, i think that any solution, when it gracefully removed, should revert the system to a clean state, just like before it was installed.

I agree with this completely in principle. In practice, for this particular case, it's hard to do so without sacrificing the correct behavior in the very real mainline use-case of upgrading Calico on a cluster.

One thing that is a remote possibility is adding code to our helm chart to handle this. It would only work when installing / uninstalling using helm, but integration with purpose built tools for managing software give us a hook to tell if Calico is being uninstalled or if it's just being upgraded that we don't otherwise have visibility to. I still think regardless of any workarounds we cna add Calico, best practice is going to be to `kubectl delete` your nodes and reboot them to ensure a clean state both in the Kubernetes API and the nodes themselves if you want a fresh slate.

**Comment by rshiva777** [comment]
for me, calico installation delete got stuck without being deleted. i tried to describe and check the logs but not getting any clues. Any lead would be helpful

kubectl delete installation default

**Comment by caseydavenport** [comment]
@rshiva777 please open a separate issue. That sounds like a different problem.

**Comment by dberardo-com** [comment]
Activity
